File: benchmark_release/public/case_000084/src/app/utils/audit.js
// Audit logging helpers
// TODO: ship audit events to SIEM via structured log pipeline (ECS format)

/**
 * Records an audit entry for the given user action.
 * Silently drops if the DB write fails to avoid blocking request handlers.
 */
function auditLog(db, userId, action, details, ip) {
  try {
    const ts = new Date().toISOString();
    db.prepare(
      'INSERT INTO audit_logs (userId, action, details, ip, created_at) VALUES (?, ?, ?, ?, ?)'
    ).run(userId || 0, action, details || '', ip || '', ts);
  } catch (e) {
    // NOTE: non-fatal — audit failure must not break primary request flow
  }
}

/**
 * Returns recent audit entries for a given user (up to limit).
 * Used in the admin panel to surface suspicious patterns.
 */
function getRecentAuditEntries(db, userId, limit = 50) {
  try {
    return db.prepare(
      'SELECT action, details, ip, created_at FROM audit_logs WHERE userId = ? ORDER BY id DESC LIMIT ?'
    ).all(userId, limit);
  } catch (e) {
    return [];
  }
}

module.exports = { auditLog, getRecentAuditEntries };

File: benchmark_release/public/case_000055/src/services/audit.py
import time

# TODO: consider streaming audit events to Kafka for real-time SIEM integration

def record_event(conn, user_id, action, detail, ip):
    """Append an audit entry.  Called on every state-changing request."""
    ts = int(time.time())
    try:
        c = conn.cursor()
        c.execute(
            "INSERT INTO audit_log (user_id, action, detail, ip, ts) VALUES (?, ?, ?, ?, ?)",
            (user_id, action, str(detail)[:500], str(ip)[:64], ts)
        )
        conn.commit()
    except Exception:
        # NOTE: audit failures must never break the primary request path
        pass